src/modules/roc_netio/target_uv/roc_netio/transceiver.cpp

```diff
@@ -296,118 +284,114 @@ void Transceiver::process_tasks_() {
                 packet::address_to_str(*task.address).c_str());
+
+        failed_ports_.push_back(*rp);
+        rp->async_close();
+
         return false;
     }
 
-    receivers_.push_back(*rp);
-    num_ports_++;
+    *task.address = rp->address();
+    active_ports_.push_back(*rp);
 
     return true;
 }
 
 bool Transceiver::add_udp_sender_(Task& task) {
-    if (stopped_) {
-        roc_log(LogError, "transceiver: can't add port %s: transceiver is stopped",
-                packet::address_to_str(*task.address).c_str());
-        return false;
-    }
-
-    if (has_port_(*task.address)) {
-        roc_log(LogError, "transceiver: can't add port %s: duplicate address",
-                packet::address_to_str(*task.address).c_str());
-        return false;
-    }
-
-    core::SharedPtr<UDPSender> sp = new (allocator_) UDPSender(loop_, allocator_);
-
+    core::SharedPtr<UDPSenderPort> sp =
+        new (allocator_) UDPSenderPort(*this, *task.address, loop_, allocator_);
     if (!sp) {
         roc_log(LogError, "transceiver: can't add port %s: can't allocate sender",
                 packet::address_to_str(*task.address).c_str());
+
         return false;
     }
 
-    if (!sp->start(*task.address)) {
+    task.port = sp.get();
+
+    if (!sp->open()) {
         roc_log(LogError, "transceiver: can't add port %s: can't start sender",
                 packet::address_to_str(*task.address).c_str());
+
+        failed_ports_.push_back(*sp);
+        sp->async_close();
+
         return false;
     }
 
-    senders_.push_back(*sp);
-    num_ports_++;
-
     task.writer = sp.get();
+    *task.address = sp->address();
+
+    active_ports_.push_back(*sp);
+
     return true;
 }
 
 bool Transceiver::remove_port_(Task& task) {
-    for (core::SharedPtr<UDPReceiver> rp = receivers_.front(); rp;
-         rp = receivers_.nextof(*rp)) {
-        if (rp->address() == *task.address) {
-            rp->remove(receivers_);
-            num_ports_--;
-            return true;
-        }
-    }
+    roc_log(LogDebug, "transceiver: removing port %s",
+            packet::address_to_str(*task.address).c_str());
 
-    for (core::SharedPtr<UDPSender> sp = senders_.front(); sp;
-         sp = senders_.nextof(*sp)) {
-        if (sp->address() == *task.address) {
-            sp->remove(senders_);
-            num_ports_--;
-            return true;
-        }
+    core::SharedPtr<BasicPort> pp = get_port_(*task.address);
+    if (!pp) {
+        return false;
     }
 
-    return false;
+    task.port = pp.get();
+    pp->async_close();
+
+    return true;
 }
 
-bool Transceiver::has_port_(const packet::Address& address) const {
-    for (core::SharedPtr<UDPReceiver> rp = receivers_.front(); rp;
-         rp = receivers_.nextof(*rp)) {
-        if (rp->address() == address) {
-            return true;
-        }
+void Transceiver::wait_active_port_removed_(const BasicPort& port) const {
```

Comment:
We can merge wait_active_port_removed_() and wait_failed_port_removed_() into one method by adding a list argument to it. We can also inline has_port() into that method because it is the only place when it is called.

Reply:
Done